source/endpoints_ruleset.c

```diff
@@ -0,0 +1,939 @@
+    } else {
+        parameter->is_deprecated = false;
+        parameter->deprecated_message = NULL;
+        parameter->deprecated_since = NULL;
+    }
+
+    aws_hash_table_put(wrapper->table, &parameter->name_cur, parameter, NULL);
+    return AWS_OP_SUCCESS;
+}
+
+static int s_on_condition_element(
+    size_t idx,
+    const struct aws_json_value *condition_node,
+    bool *out_should_continue,
+    void *user_data) {
+    (void)idx;
+    (void)out_should_continue;
+    struct array_parser_wrapper *wrapper = user_data;
+
+    struct aws_endpoints_condition condition;
+    AWS_ZERO_STRUCT(condition);
+
+    if (s_parse_function(wrapper->allocator, condition_node, &condition.function)) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to parse function.");
+        return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+    }
+
+    struct aws_json_value *assign_node =
+        aws_json_value_get_from_object(condition_node, aws_byte_cursor_from_c_str("assign"));
+    if (assign_node != NULL) {
+        struct aws_byte_cursor cur;
+        if (aws_json_value_get_string(assign_node, &cur)) {
+            aws_endpoints_condition_clean_up(&condition);
+            AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Unexpected value for assign.");
+            return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+        }
+        condition.assign = aws_string_new_from_cursor(wrapper->allocator, &cur);
+    }
+
+    aws_array_list_push_back(wrapper->array, &condition);
+    return AWS_OP_SUCCESS;
+}
+
+static int s_on_header_element(
+    size_t idx,
+    const struct aws_json_value *value,
+    bool *out_should_continue,
+    void *user_data) {
+    (void)idx;
+    (void)out_should_continue;
+    struct array_parser_wrapper *wrapper = user_data;
+    struct aws_byte_cursor cur;
+    aws_json_value_get_string(value, &cur);
+
+    struct aws_string *string = aws_string_new_from_cursor(wrapper->allocator, &cur);
+    aws_array_list_push_back(wrapper->array, &string);
+    return true;
+}
+
+static int s_on_headers_key(
+    const struct aws_byte_cursor *key,
+    const struct aws_json_value *value,
+    bool *out_should_continue,
+    void *user_data) {
+    (void)out_should_continue;
+    struct member_parser_wrapper *wrapper = user_data;
+
+    if (aws_json_value_is_array(value)) {
+        size_t num_elements = aws_json_get_array_size(value);
+        struct aws_array_list *headers = aws_mem_calloc(wrapper->allocator, 1, sizeof(struct aws_array_list));
+        aws_array_list_init_dynamic(headers, wrapper->allocator, num_elements, sizeof(struct aws_string *));
+        if (s_init_array_from_json(wrapper->allocator, value, headers, s_on_header_element)) {
+            AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to extract url.");
+            return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
```

Comment:
failing to clean up `headers` in the case of error.

I'd suggest something like :
```
static int s_on_headers_key(...args...) {
    struct aws_array_list *headers = NULL;
    
    ... rest of code if anything goes wrong then goto error;

    return AWS_OP_SUCCESS;
error:
    if (headers) {
        aws_array_list_deep_cleanup(...)
        aws_mem_release(headers);
    }
    return AWS_OP_ERR;
}

```